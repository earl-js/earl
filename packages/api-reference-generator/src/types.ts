export interface MethodComment {
  signature: string
  comment: string
}

export interface MethodDocumentation {
  signature: string
  description: string
  params: Param[]
  examples: string[]
}

export interface Param {
  name: string
  description: string
}
