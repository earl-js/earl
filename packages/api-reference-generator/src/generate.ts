import { sortBy } from 'lodash'

import { MethodDocumentation } from './types'

export function generateMarkdownForMethodDocumentation(doc: MethodDocumentation): string {
  return `#### ${doc.signature}

${doc.description}

*Parameters:*

${doc.params.map((p) => `- \`${p.name}\` - ${p.description}`).join('\n')}

*Examples:*

${doc.examples.join('\n')}`
}

export function generateTableOfContents(docs: MethodDocumentation[]) {
  const links = docs.map((d) => `- [${encodeLabel(d.signature)}](#${encodeAnchor(d.signature)})`)

  return links.join('\n\n')
}

// I was unable to find a ready implementation for this so i ended up writing my own half-assed implementation. This might need tweaking...
function encodeAnchor(input: string): string {
  return input
    .toLowerCase()
    .replace(/[(),:]/g, '')
    .replace(/ /g, '-')
}

function encodeLabel(input: string): string {
  return input.replace(/</g, '\\<')
}
