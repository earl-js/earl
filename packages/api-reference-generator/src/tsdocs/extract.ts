import { assert } from 'ts-essentials'

import { MethodComment } from '../types'

export function extractTsDocCommentsFromString(source: string): MethodComment[] {
  const BLOCK_COMMENT_REGEX = /\/\*\*(.*?)\*\/\n(.*?)[{|\n]/gms

  const methodComments: MethodComment[] = []

  let rawMethodComment = BLOCK_COMMENT_REGEX.exec(source)
  assert(rawMethodComment, `Couldn't find any block comments in ${source}`)
  while (rawMethodComment != null) {
    const comment = `/** ${rawMethodComment[1].trim()} */`
    const signature = removeGetterKeyword(rawMethodComment[2].trim())

    methodComments.push({ signature, comment })

    rawMethodComment = BLOCK_COMMENT_REGEX.exec(source)
  }

  return methodComments
}

function removeGetterKeyword(signature: string): string {
  if (signature.startsWith('get ')) {
    return signature.substring('get '.length)
  }
  return signature
}
