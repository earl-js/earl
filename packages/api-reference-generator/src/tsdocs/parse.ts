import { DocExcerpt, DocNode, TSDocParser } from '@microsoft/tsdoc'

import { MethodComment, MethodDocumentation, Param } from '../types'

export function parseTsDocComment(methodComment: MethodComment): MethodDocumentation {
  const tsdocParser: TSDocParser = new TSDocParser()

  const parserContext = tsdocParser.parseString(methodComment.comment)

  if (parserContext.log.messages.length > 0) {
    throw new Error('Syntax error: ' + parserContext.log.messages[0].text)
  }

  const docComment = parserContext.docComment

  const description = Formatter.renderDocNode(docComment.summarySection).trimRight()

  const params: Param[] = []
  for (const param of docComment.params.blocks) {
    params.push({ name: param.parameterName, description: Formatter.renderDocNode(param.content).trim() })
  }

  const examples: string[] = []

  for (const customBlock of docComment.customBlocks) {
    if (customBlock.blockTag.tagName !== '@example') {
      continue
    }

    const contents = Formatter.renderDocNode(customBlock.content).trim()
    examples.push(contents)
  }

  return {
    signature: methodComment.signature,
    description,
    params,
    examples,
  }
}

class Formatter {
  public static renderDocNode(docNode: DocNode): string {
    let result: string = ''

    if (docNode instanceof DocExcerpt) {
      result += docNode.content.toString()
    }
    for (const childNode of docNode.getChildNodes()) {
      result += Formatter.renderDocNode(childNode)
    }

    return result
  }

  public static renderDocNodes(docNodes: ReadonlyArray<DocNode>): string {
    let result: string = ''
    for (const docNode of docNodes) {
      result += Formatter.renderDocNode(docNode)
    }
    return result
  }
}
