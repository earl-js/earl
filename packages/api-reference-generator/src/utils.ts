// I was unable to find a ready implementation for this so i ended up writing my own half-assed implementation. This might need tweaking...
export function encodeAnchor(input: string): string {
  return input
    .toLowerCase()
    .replace(/[()\.,:\[\]]/g, '')
    .replace(/<(.*?)>/g, '')
    .replace(/ /g, '-')
}

export function encodeLabel(input: string): string {
  return input.replace(/</g, '\\<')
}

znajdz algorytm odpowiedzialny za generarowanie linkow inaczej to lipa