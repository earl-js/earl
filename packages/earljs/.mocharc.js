// ensure NODE_ENV
process.env.NODE_ENV = 'test'

// exit test runner on unhandled rejections
process.on('unhandledRejection', (reason, promise) => {
  console.error('Unhandled Rejection during test execution:', promise, 'reason:', reason)
  process.exit(1)
})

module.exports = {
  require: ['ts-node/register/transpile-only', './test/setup.ts', './src/test-runners/mocha.ts'],
  extension: 'ts',
  watchExtensions: 'ts',
  spec: 'test/**/*.test.ts',
}
