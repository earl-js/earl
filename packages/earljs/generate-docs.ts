import { TSDocParser, ParserContext, DocNode, DocExcerpt } from '@microsoft/tsdoc'
import { readFileSync } from 'fs'
import { join } from 'path'
import { assert } from 'ts-essentials'

interface Param {
  name: string
  description: string
}

interface Example {
  description: string
  source: string
}

interface MethodDocumentation {
  signature: string
  description: string
  params: Param[]
  examples: Example[]
}

export function parseComment(methodComment: MethodComment): MethodDocumentation {
  const tsdocParser: TSDocParser = new TSDocParser()

  const parserContext = tsdocParser.parseString(methodComment.comment)

  if (parserContext.log.messages.length > 0) {
    throw new Error('Syntax error: ' + parserContext.log.messages[0].text)
  }

  const docComment = parserContext.docComment

  const description = Formatter.renderDocNode(docComment.summarySection)
  const params: Param[] = []
  for (const param of docComment.params.blocks) {
    params.push({ name: param.parameterName, description: Formatter.renderDocNode(param.content).trim() })
  }

  return {
    signature: methodComment.signature,
    description,
    params,
    examples: [],
  }
}

export interface MethodComment {
  signature: string
  comment: string
}

export function extractTsDocCommentsFromString(source: string): MethodComment[] {
  const BLOCK_COMMENT_REGEX = /\/\*\*(.*?)\*\/\n(.*?)[\{|\n]/gms

  const methodComments: MethodComment[] = []

  let rawMethodComment = BLOCK_COMMENT_REGEX.exec(source)
  assert(rawMethodComment, `Couldn't find any block comments in ${source}`)
  while (rawMethodComment != null) {
    debugger
    const comment = `/** ${rawMethodComment[1].trim()} */`
    const signature = rawMethodComment[2].trim()

    methodComments.push({ signature, comment })

    rawMethodComment = BLOCK_COMMENT_REGEX.exec(source)
  }

  return methodComments
}

export class Formatter {
  public static renderDocNode(docNode: DocNode): string {
    let result: string = ''
    if (docNode) {
      if (docNode instanceof DocExcerpt) {
        result += docNode.content.toString()
      }
      for (const childNode of docNode.getChildNodes()) {
        result += Formatter.renderDocNode(childNode)
      }
    }
    return result
  }

  public static renderDocNodes(docNodes: ReadonlyArray<DocNode>): string {
    let result: string = ''
    for (const docNode of docNodes) {
      result += Formatter.renderDocNode(docNode)
    }
    return result
  }
}
