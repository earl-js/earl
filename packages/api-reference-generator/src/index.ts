import { readFileSync, writeFileSync } from 'fs'
import { sortBy } from 'lodash'
import { join } from 'path'

import { generateMarkdownForMethodDocumentation, generateTableOfContents } from './generate'
import { extractTsDocCommentsFromString } from './tsdocs/extract'
import { parseTsDocComment } from './tsdocs/parse'

export function generateApiReference() {
  const basePath = join(__dirname, '../../earljs/src')

  const files = {
    Validators: join(basePath, './Expectation.ts'),
  }

  const source = readFileSync(files.Validators, 'utf-8')
  const comments = extractTsDocCommentsFromString(source)
  const parsed = comments.map(parseTsDocComment)
  const sorted = sortBy(parsed, (d) => d.signature)

  const output = `
## Validators
${generateTableOfContents(sorted)}

# Reference

## Validators
${sorted.map(generateMarkdownForMethodDocumentation).join('\n')}
  `

  writeFileSync(join(__dirname, './out.md'), output, 'utf-8')
  console.log('DONE!')
}

generateApiReference()
