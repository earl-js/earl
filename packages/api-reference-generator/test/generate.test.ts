import { expect } from 'earljs'

import { generateMarkdownForMethodDocumentation } from '../src/generate'
import { parseTsDocComment } from '../src/tsdocs/parse'
import { sampleMethodComment } from './tsdocs/parse.test'

describe('generateMarkdownForMethodDocumentation', () => {
  it('generates markdown for a comment with params and examples', async () => {
    const result = generateMarkdownForMethodDocumentation(parseTsDocComment(sampleMethodComment))

    expect(result).toMatchSnapshot()
  })
})
