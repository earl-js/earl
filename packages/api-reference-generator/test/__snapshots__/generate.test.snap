// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMarkdownForMethodDocumentation generates markdown for a comment with params and examples 1`] = `
"#### someMethod(x: number, y: number): void

Returns the average of two numbers.

Some other note

*Parameters:*

- \`x\` - The first input number
- \`y\` - The second input number

*Examples:*

Random example
\`\`\`ts
someMethod(1, 2)
\`\`\`"
`;

exports[`generateTableOfContents generates valid TOC 1`] = `
"- [sample()](#sample)

- [expect(value: any, options: Options)](#expectvalue-any-options-options)

- [expect\\\\<TYPE>(value: any, options: Options)](#expect<type>value-any-options-options)"
`;
